ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -175,70 +183,140 @@ public static TableRowsEntity execute(
+        "Unable to execute pull query: %s", statement.getStatementText()));
+  }
 
-      return new TableRowsEntity(
-          statement.getStatementText(),
-          queryId,
+  private TableRowsEntity routeQuery(
+      final KsqlNode node,
+      final ConfiguredStatement<Query> statement,
+      final KsqlExecutionContext executionContext,
+      final ServiceContext serviceContext,
+      final PullQueryContext pullQueryContext
+  ) {
+
+    if (node.isLocal()) {
+      LOG.debug("Query {} executed locally at host {} at timestamp {}.",
+               statement.getStatementText(), node.location(), System.currentTimeMillis());
+      return queryRowsLocally(
+          statement,
+          executionContext,
+          pullQueryContext);
+    } else {
+      LOG.debug("Query {} routed to host {} at timestamp {}.",
+                statement.getStatementText(), node.location(), System.currentTimeMillis());
+      return forwardTo(node, statement, serviceContext);
+    }
+  }
+
+  @VisibleForTesting
+  TableRowsEntity queryRowsLocally(
+      final ConfiguredStatement<Query> statement,
+      final KsqlExecutionContext executionContext,
+      final PullQueryContext pullQueryContext
+  ) {
+    final Result result;
+    if (pullQueryContext.whereInfo.windowStartBounds.isPresent()) {
+      final Range<Instant> windowStart = pullQueryContext.whereInfo.windowStartBounds.get();
+
+      final List<? extends TableRow> rows = pullQueryContext.mat.windowed()
+          .get(pullQueryContext.rowKey, windowStart);
+
+      result = new Result(pullQueryContext.mat.schema(), rows);
+    } else {
+      final List<? extends TableRow> rows = pullQueryContext.mat.nonWindowed()
+          .get(pullQueryContext.rowKey)
+          .map(ImmutableList::of)
+          .orElse(ImmutableList.of());
+
+      result = new Result(pullQueryContext.mat.schema(), rows);
+    }
+
+    final LogicalSchema outputSchema;
+    final List<List<?>> rows;
+    if (isSelectStar(statement.getStatement().getSelect())) {
+      outputSchema = TableRowsEntityFactory.buildSchema(
+          result.schema, pullQueryContext.mat.windowType().isPresent());
+      rows = TableRowsEntityFactory.createRows(result.rows);
+    } else {
+      outputSchema = selectOutputSchema(
+          result, executionContext, pullQueryContext.analysis, pullQueryContext.mat.windowType());
+      rows = handleSelects(
+          result,
+          statement,
+          executionContext,
+          pullQueryContext.analysis,
           outputSchema,
-          rows
-      );
-    } catch (final Exception e) {
-      throw new KsqlStatementException(
-          e.getMessage() == null ? "Server Error" : e.getMessage(),
-          statement.getStatementText(),
-          e
+          pullQueryContext.mat.windowType(),
+          pullQueryContext.queryId,
+          pullQueryContext.contextStacker
       );
     }
+    System.out.println("Result of pull query = " + rows);
```

Comment:
remove or move to logger.debug??

Reply:
Sorry, intellij indexing doesn't work. I have searched to remove all `System.out` but nothing came up